server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1811,6 +1883,51 @@ public void finalizeSnapshot(
         }, onUpdateFailure);
     }
 
+    private void cleanRemoteStoreDirectoryIfNeeded(
+        String indexSnId,
+        RepositoryData oldRepoData,
+        RemoteSegmentStoreDirectoryFactory remoteSegmentStoreDirectoryFactory
+    ) {
+        IndexId indexId = null;
+        List<SnapshotId> snapshotIds = Collections.emptyList();
+        for (Map.Entry<IndexId, List<SnapshotId>> entry : oldRepoData.getIndexSnapshots().entrySet()) {
+            indexId = entry.getKey();
+            if (indexId.getId().equals(indexSnId)) {
+                snapshotIds = entry.getValue();
+                break;
+            }
+        }
+
+        for (SnapshotId snapshotId : snapshotIds) {
+            try {
+                IndexMetadata prevIndexMetadata = this.getSnapshotIndexMetaData(oldRepoData, snapshotId, indexId);
+                if (prevIndexMetadata != null && !isIndexPresent(clusterService, prevIndexMetadata.getIndexUUID())) {
+                    String remoteStoreRepository = IndexMetadata.INDEX_REMOTE_TRANSLOG_REPOSITORY_SETTING.get(
+                        prevIndexMetadata.getSettings()
+                    );
+                    if (remoteStoreRepository == null) {
+                        break;
```

Comment:
should this be an assert or `IllegalArgumentException`. ?

Reply:
Makes sense, since the code path is executed only for shallow snapshots. Added an assert